casper/src/main/scala/coop/rchain/casper/Validate.scala

```diff
@@ -312,13 +312,17 @@ object Validate {
     } yield result
 
   // Agnostic of non-parent justifications
-  def blockNumber[F[_]: Monad: Log: BlockStore](
-      b: BlockMessage
+  def blockNumber[F[_]: Monad: Log](
+      b: BlockMessage,
+      dag: BlockDagRepresentation[F]
   ): F[Either[InvalidBlock, ValidBlock]] =
     for {
-      parents <- ProtoUtil.unsafeGetParents[F](b)
+      maybeParents <- ProtoUtil.parentHashes(b).toList.traverse { parentHash =>
+                       dag.lookup(parentHash)
+                     }
+      parents = maybeParents.flatten
```

Comment:
Are you sure you want to flatten `List[Option[BlockMetadata]]` here? I think all of metadatas should be present here, so maybe just check if they are indeed there and raise error otherwise?

Reply:
As in just `parents = maybeParents.map(_.get)`? We seem to have this issue all over the codebase: we sometimes handle it with just .get, and sometimes throw an error - I'm not sure what is the best path forward here.

@rabbitonweb Any ideas?

Reply:
More like as in:
```scala
maybeParents.sequence[Option, BlockMetadata] match {
  case Some(parents) => ...
  case None => Sync[F].raise(InvalidBlockDagStorageState(...))
}
```

Reply:
Ah ok; I like that